### PR TITLE
Refactor : the NAT Gateway resource and data source to use the framework

### LINKED ISSUE #341: Refactor the NAT Gateway resource and data source to use the framework
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Refactoring the `NAT Gateway` resource, data source the Framework

### New or Affected Resource(s)

* ncloud_nat_gateway_*

### Potential Terraform Configuration
```
None
```

### References
- [ncloud nat gateway docs](https://api.ncloud-docs.com/docs/networking-vpc-natgateway-createnatgatewayinstance)

### PR DESCRIPTION
**Pull request by royroyee**
### Description
resolve #341 

refactor the `Nat Gateway` resource and data source to use the framework

#### Resource
- Nat Gateway

#### DataSource
- Nat Gateway

### Test
```
go test -run=TestAccResourceNcloudNatGateway_basic
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc  

go test -run=TestAccResourceNcloudNatGateway_disappears
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc 

go test -run=TestAccResourceNcloudNatGateway_onlyRequiredParam
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc

go test -run=TestAccResourceNcloudNatGateway_updateName
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc

go test -run=TestAccResourceNcloudNatGateway_description
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc 

// DataSource
go test -run=TestAccDataSourceNcloudNatGateway_basic
PASS
ok      github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc 
```

#### SDKv2 Nat Gateway Resource State
```
# ncloud_nat_gateway.nat_gateway:
resource "ncloud_nat_gateway" "nat_gateway" {
    description    = "test"
    id             = "19002129"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002129"
    private_ip     = "10.3.1.6"
    public_ip      = "175.106.96.159"
    public_ip_no   = "19002146"
    subnet_name    = "sn18a034ce137"
    subnet_no      = "99389"
    vpc_no         = "44527"
    zone           = "KR-1"
}

# ncloud_nat_gateway.nat_gateway_private:
resource "ncloud_nat_gateway" "nat_gateway_private" {
    description    = "test"
    id             = "19002128"
    name           = "ng18a034d0e6a"
    nat_gateway_no = "19002128"
    private_ip     = "10.3.2.6"
    subnet_name    = "sn18a034ce176"
    subnet_no      = "99390"
    vpc_no         = "44527"
    zone           = "KR-1"
}

# ncloud_subnet.subnet_private:
resource "ncloud_subnet" "subnet_private" {
    id             = "99390"
    name           = "sn18a034ce176"
    network_acl_no = "68897"
    subnet         = "10.3.2.0/24"
    subnet_no      = "99390"
    subnet_type    = "PRIVATE"
    usage_type     = "NATGW"
    vpc_no         = "44527"
    zone           = "KR-1"
}

# ncloud_subnet.subnet_public:
resource "ncloud_subnet" "subnet_public" {
    id             = "99389"
    name           = "sn18a034ce137"
    network_acl_no = "68897"
    subnet         = "10.3.1.0/24"
    subnet_no      = "99389"
    subnet_type    = "PUBLIC"
    usage_type     = "NATGW"
    vpc_no         = "44527"
    zone           = "KR-1"
}

# ncloud_vpc.vpc:
resource "ncloud_vpc" "vpc" {
    default_access_control_group_no = "123555"
    default_network_acl_no          = "68897"
    default_private_route_table_no  = "93758"
    default_public_route_table_no   = "93757"
    id                              = "44527"
    ipv4_cidr_block                 = "10.3.0.0/16"
    name                            = "test-nat-gateway"
    vpc_no                          = "44527"
}
```

#### Framework Nat Gateway Resource State
```
# ncloud_nat_gateway.nat_gateway:
resource "ncloud_nat_gateway" "nat_gateway" {
    description    = "test"
    id             = "19001953"
    name           = "test-nat-gateway"
    nat_gateway_no = "19001953"
    private_ip     = "10.3.1.6"
    public_ip      = "223.130.129.106"
    public_ip_no   = "19001958"
    subnet_name    = "sn18a03451466"
    subnet_no      = "99385"
    vpc_no         = "44523"
    zone           = "KR-1"
}

# ncloud_nat_gateway.nat_gateway_private:
resource "ncloud_nat_gateway" "nat_gateway_private" {
    description    = "test"
    id             = "19001952"
    name           = "ng18a03454152"
    nat_gateway_no = "19001952"
    private_ip     = "10.3.2.6"
    subnet_name    = "sn18a03451454"
    subnet_no      = "99384"
    vpc_no         = "44523"
    zone           = "KR-1"
}

# ncloud_subnet.subnet_private:
resource "ncloud_subnet" "subnet_private" {
    id             = "99384"
    name           = "sn18a03451454"
    network_acl_no = "68893"
    subnet         = "10.3.2.0/24"
    subnet_no      = "99384"
    subnet_type    = "PRIVATE"
    usage_type     = "NATGW"
    vpc_no         = "44523"
    zone           = "KR-1"
}

# ncloud_subnet.subnet_public:
resource "ncloud_subnet" "subnet_public" {
    id             = "99385"
    name           = "sn18a03451466"
    network_acl_no = "68893"
    subnet         = "10.3.1.0/24"
    subnet_no      = "99385"
    subnet_type    = "PUBLIC"
    usage_type     = "NATGW"
    vpc_no         = "44523"
    zone           = "KR-1"
}

# ncloud_vpc.vpc:
resource "ncloud_vpc" "vpc" {
    default_access_control_group_no = "123544"
    default_network_acl_no          = "68893"
    default_private_route_table_no  = "93750"
    default_public_route_table_no   = "93749"
    id                              = "44523"
    ipv4_cidr_block                 = "10.3.0.0/16"
    name                            = "test-nat-gateway"
    vpc_no                          = "44523"
}
```

#### SDKv2 Nat Gateway DataSource State
```
# data.ncloud_nat_gateway.by_filter:
data "ncloud_nat_gateway" "by_filter" {
    description    = "description"
    id             = "19002095"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002095"
    private_ip     = "10.3.1.6"
    public_ip      = "110.165.17.91"
    public_ip_no   = "19002097"
    subnet_name    = "sn18a034a7398"
    subnet_no      = "99388"
    vpc_name       = "test-nat-gateway"
    vpc_no         = "44526"
    zone           = "KR-1"

    filter {
        name   = "nat_gateway_no"
        regex  = false
        values = [
            "19002095",
        ]
    }
}

# data.ncloud_nat_gateway.by_id:
data "ncloud_nat_gateway" "by_id" {
    description    = "description"
    id             = "19002095"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002095"
    private_ip     = "10.3.1.6"
    public_ip      = "110.165.17.91"
    public_ip_no   = "19002097"
    subnet_name    = "sn18a034a7398"
    subnet_no      = "99388"
    vpc_name       = "test-nat-gateway"
    vpc_no         = "44526"
    zone           = "KR-1"
}

# ncloud_nat_gateway.nat_gateway:
resource "ncloud_nat_gateway" "nat_gateway" {
    description    = "description"
    id             = "19002095"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002095"
    private_ip     = "10.3.1.6"
    public_ip      = "110.165.17.91"
    public_ip_no   = "19002097"
    subnet_name    = "sn18a034a7398"
    subnet_no      = "99388"
    vpc_no         = "44526"
    zone           = "KR-1"
}

# ncloud_subnet.subnet:
resource "ncloud_subnet" "subnet" {
    id             = "99388"
    name           = "sn18a034a7398"
    network_acl_no = "68896"
    subnet         = "10.3.1.0/24"
    subnet_no      = "99388"
    subnet_type    = "PUBLIC"
    usage_type     = "NATGW"
    vpc_no         = "44526"
    zone           = "KR-1"
}

# ncloud_vpc.vpc:
resource "ncloud_vpc" "vpc" {
    default_access_control_group_no = "123553"
    default_network_acl_no          = "68896"
    default_private_route_table_no  = "93756"
    default_public_route_table_no   = "93755"
    id                              = "44526"
    ipv4_cidr_block                 = "10.3.0.0/16"
    name                            = "test-nat-gateway"
    vpc_no                          = "44526"
}
```

#### Framework Nat Gateway DataSource State
```
# data.ncloud_nat_gateway.by_filter:
data "ncloud_nat_gateway" "by_filter" {
    description    = "description"
    id             = "19002044"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002044"
    private_ip     = "10.3.1.6"
    public_ip      = "223.130.130.123"
    public_ip_no   = "19002045"
    subnet_name    = "sn18a0347b154"
    subnet_no      = "99386"
    vpc_name       = "test-nat-gateway"
    vpc_no         = "44524"
    zone           = "KR-1"

    filter {
        name   = "nat_gateway_no"
        values = [
            "19002044",
        ]
    }
}

# data.ncloud_nat_gateway.by_id:
data "ncloud_nat_gateway" "by_id" {
    description    = "description"
    id             = "19002044"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002044"
    private_ip     = "10.3.1.6"
    public_ip      = "223.130.130.123"
    public_ip_no   = "19002045"
    subnet_name    = "sn18a0347b154"
    subnet_no      = "99386"
    vpc_name       = "test-nat-gateway"
    vpc_no         = "44524"
    zone           = "KR-1"
}

# ncloud_nat_gateway.nat_gateway:
resource "ncloud_nat_gateway" "nat_gateway" {
    description    = "description"
    id             = "19002044"
    name           = "test-nat-gateway"
    nat_gateway_no = "19002044"
    private_ip     = "10.3.1.6"
    public_ip      = "223.130.130.123"
    public_ip_no   = "19002045"
    subnet_name    = "sn18a0347b154"
    subnet_no      = "99386"
    vpc_no         = "44524"
    zone           = "KR-1"
}

# ncloud_subnet.subnet:
resource "ncloud_subnet" "subnet" {
    id             = "99386"
    name           = "sn18a0347b154"
    network_acl_no = "68894"
    subnet         = "10.3.1.0/24"
    subnet_no      = "99386"
    subnet_type    = "PUBLIC"
    usage_type     = "NATGW"
    vpc_no         = "44524"
    zone           = "KR-1"
}

# ncloud_vpc.vpc:
resource "ncloud_vpc" "vpc" {
    default_access_control_group_no = "123550"
    default_network_acl_no          = "68894"
    default_private_route_table_no  = "93752"
    default_public_route_table_no   = "93751"
    id                              = "44524"
    ipv4_cidr_block                 = "10.3.0.0/16"
    name                            = "test-nat-gateway"
    vpc_no                          = "44524"
}
```

close #341

### CODE REVIEW

File: internal/service/vpc/nat_gateway.go [hunk 1]
```diff
@@ -1,205 +1,320 @@
...
-	d.Set("vpc_no", instance.VpcNo)
-	d.Set("zone", instance.ZoneCode)
-	d.Set("subnet_name", instance.SubnetName)
-	d.Set("subnet_no", instance.SubnetNo)
-	d.Set("private_ip", instance.PrivateIp)
-	d.Set("public_ip_no", instance.PublicIpInstanceNo)
+	if err := state.refreshFromOutput(output); err != nil {
+		resp.Diagnostics.AddError("refreshing nat gateway details", err.Error())
+	}
+	state.Description = framework.EmptyStringToNull(ctx, state.Description)
```
Comment (wonchulee) [inline internal/service/vpc/nat_gateway.go]: 사소한 부분이지만, Create,Read,Update에서 공통으로 사용되니 EmptyStringToNull 로 null 값 변경해주는 부분을 refreshFromOutput 내에서 함께 수행해주면 깔끔할 것 같습니다

**Comment by wonchulee** [comment]
All the tests are passed after applying the framework change.

=== RUN   TestAccResourceNcloudNatGateway_basic
--- PASS: TestAccResourceNcloudNatGateway_basic (92.47s)
=== RUN   TestAccResourceNcloudNatGateway_disappears
--- PASS: TestAccResourceNcloudNatGateway_disappears (111.66s)
=== RUN   TestAccResourceNcloudNatGateway_onlyRequiredParam
--- PASS: TestAccResourceNcloudNatGateway_onlyRequiredParam (90.29s)
=== RUN   TestAccResourceNcloudNatGateway_updateName
--- PASS: TestAccResourceNcloudNatGateway_updateName (90.52s)
=== RUN   TestAccResourceNcloudNatGateway_description
--- PASS: TestAccResourceNcloudNatGateway_description (145.27s)
PASS
ok  	github.com/terraform-providers/terraform-provider-ncloud/internal/service/vpc	530.787s

File: internal/service/vpc/nat_gateway.go (hunk 1 above)
Comment (royroyee) [reply internal/service/vpc/nat_gateway.go]: 아 그렇네요 공통되는 부분이었네요.

그러면 아래와 같이 refreshFromOutput 에 추가하면 괜찮을까요? Description 부분만 한 번 더 체크하도록 했습니다.

```go
func (m *natGatewayResourceModel) refreshFromOutput(output *vpc.NatGatewayInstance) error {
	m.ID = types.StringPointerValue(output.NatGatewayInstanceNo)
	m.NatGatewayNo = types.StringPointerValue(output.NatGatewayInstanceNo)
	m.Name = types.StringPointerValue(output.NatGatewayName)
	m.Description = types.StringPointerValue(output.NatGatewayDescription)
	m.VpcNo = types.StringPointerValue(output.VpcNo)
	m.Zone = types.StringPointerValue(output.ZoneCode)
	m.SubnetNo = types.StringPointerValue(output.SubnetNo)
	m.PrivateIp = types.StringPointerValue(output.PrivateIp)
	m.PublicIpNo = types.StringPointerValue(output.PublicIpInstanceNo)
	m.PublicIp = types.StringPointerValue(output.PublicIp)
	m.SubnetName = types.StringPointerValue(output.SubnetName)

	m.Description = framework.EmptyStringToNull(m.Description)

	return nil
}
```

수정 이후에도 테스트는 모두 통과했습니다

Comment (wonchulee) [reply internal/service/vpc/nat_gateway.go]: 넵,
`m.Description = framework.EmptyStringToNull(types.StringPointerValue(output.NatGatewayDescription))`  요렇게 넣어주셔도 좋을 것 같습니다.

Comment (royroyee) [reply internal/service/vpc/nat_gateway.go]: 아! 훨씬 깔끔하네요... 수정 완료했습니다.

감사합니다!

**Review by wonchulee — APPROVED** [review]

**Pull request merged**